File: 5_Recursion/1_ProductSum.cpp
#include <any>
#include <vector>
#include <cassert>

using namespace std;

// Tip: You can use el.type() == typeid(vector<any>) to check whether an item is
// a list or an integer.
// If you know an element of the array is vector<any> you can cast it using:
//     any_cast<vector<any>>(element)
// If you know an element of the array is an int you can cast it using:
//     any_cast<int>(element)
int productSum(vector<any> array,int depth = 1) {
  depth++;
  int sum = 0;
  for (auto & el : array) {
    if (el.type() == typeid(vector<any>)) {
      sum = sum + depth * productSum(std::any_cast<std::vector<any>>(el),depth);
    } else {
      sum = sum + std::any_cast<int>(el);
    }
  }
  return sum;
}


int main() {
  std::vector<std::any> myVector = {5, 2, std::vector<std::any>{7, -1}, 3, std::vector<std::any>{6, std::vector<std::any>{-13, 8}, 4}};
  assert(productSum(myVector) == 12);
}

File: 14_Recursion/1_ProductSum.py
# Tip: You can use the type(element) function to check whether an item
# is a list or an integer.

# Product Sum.[x,y] = x + y pero [x,[y,z]] -> x + 2 * (y + z)

def productSum(array):
    # Write your code here.
    return auxProduct(array)

def auxProduct(array,depth = 1):
    sum = 0
    depth = depth + 1
    prod = 1
    for i in array:
        if type(i) is int:
            sum += i

        elif type(i) is list:
            sum += depth * auxProduct(i,depth)
    return sum
        